api/client/copy.go

```diff
@@ -0,0 +1,218 @@
+package client
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/docker/docker/pkg/archive"
+)
+
+type copyDirection int
+
+const (
+	fromContainer copyDirection = (1 << iota)
+	toContainer
+	acrossContainers = fromContainer | toContainer
+)
+
+// CmdCp handles copying files or directories to/from/across containers.
+func (cli *DockerCli) CmdCp(args ...string) error {
+	cmd := cli.Subcmd("cp", "[SRC_CONTAINER:]SRC_PATH [DST_CONTAINER:]DST_PATH", "Copy files between containers or a local path", true)
+	if err := cmd.Parse(args); err != nil {
+		return nil
+	}
+
+	if cmd.NArg() != 2 {
+		cmd.Usage()
+		return nil
+	}
+
+	var (
+		srcPath      = cmd.Arg(0)
+		dstPath      = cmd.Arg(1)
+		srcContainer string
+		dstContainer string
+		direction    copyDirection
+	)
+
+	if strings.Contains(srcPath, ":") {
+		// Copy from a container.
+		direction |= fromContainer
+		srcInfo := strings.SplitN(srcPath, ":", 2)
+		srcContainer, srcPath = srcInfo[0], srcInfo[1]
+	}
+
+	if strings.Contains(dstPath, ":") {
+		// Copy to a container.
+		direction |= toContainer
+		dstInfo := strings.SplitN(dstPath, ":", 2)
+		dstContainer, dstPath = dstInfo[0], dstInfo[1]
+	}
+
+	switch direction {
+	case fromContainer:
+		// Copy from a container to a local path.
+		return cli.copyFromContainer(srcContainer, srcPath, dstPath)
+	case toContainer:
+		// Copy from a local path to a container.
+		return cli.copyToContainer(srcPath, dstContainer, dstPath)
+	case acrossContainers:
+		// Copy from one container to another.
+		return cli.copyAcrossContainers(srcContainer, srcPath, dstContainer, dstPath)
+	default:
+		// User didn't specify any container.
+		return fmt.Errorf("error: Wrong path format. Must specify a source and/or destination container")
+	}
+}
+
+type containerPathStat struct {
+	Name    string      `json:"name"`
+	AbsPath string      `json:"absPath"`
+	Size    int64       `json:"size"`
+	Mode    os.FileMode `json:"mode"`
+	Mtime   time.Time   `json:"mtime"`
+}
+
+func (cli *DockerCli) statContainerPath(containerName, path string) (containerPathStat, error) {
+	var stat containerPathStat
+
+	query := make(url.Values, 1)
+	query.Set("path", path)
+
+	urlStr := fmt.Sprintf("/containers/%s/stat-path?%s", containerName, query.Encode())
```

Comment:
I suggested that on the original PR, but there was some uncertainty if this could lead to possibly ambiguous URLs; https://github.com/docker/docker/pull/9871#issuecomment-68556640